### PR TITLE
Feat: Source Folders & Priority + Optional Alphabetical Sort

### PR DESCRIPTION
**Pull request by amirsaam**
Introduces Source Folders in IPA Library where changes the structure of Sources and adds Name and Logo URL to them and shows them as they are.
With this approach we are no longer stack all apps from all sources together so for example we can have Genshin Impact from every source in their folders.
- [x] Optimised `StoreVM`
- [x] Needed UI Changes
- [x] Working Move Source Up/Down in the Setting
- [x] Backport Compatibility for older Sources
- [x] Add Search Apps Back
- [x] Fix UI Issues
- [x] Add Alphabetical Sort Option for Source Contents
- [x] Adds collapsible list of source folders under IPA Library in sidebar

**Note**:
There is a new package `Semaphore` used in this PR, this package introduces `AsyncSemaphore` alike `DispatchSemaphore` but unlike that, it does not block any thread; Instead, Swift concurrency tasks are suspended "awaiting" for the semaphore.

This is a good thing to have for future uses like I have done in this PR dealing with suspension in the easiest way.
The reason I did not use `try await Task.sleep` is that we aren't sure when the previous item in the loop is done to start the loop again, plus we are updating UI and we need to be on main thread and `DispatchSemaphore` will not work and sends PlayCover to Not Responding state.

**Issues**:
1. When users search apps, in meantime if they change sort option it will not change the app list until adds/removes a character from their search (just saying, I don't think it's a big deal)
2.  **FIXED** | Original issue:
```
The famous Search bar issue, if gets focused and back to root it will not get unfocused until clicked on the window but still will comeback after opening a source folder again until you switch once to App Library (I have no fix)
```

### CODE REVIEW

**Comment by amirsaam** [comment]
New Source Struct + Old Struct Support:
Screenshot 1402-01-11 at 21 18 06
When opened a folder in IPA Library:
Screenshot 1402-01-11 at 21 21 33
When opened from sidebar:
Screenshot 1402-01-11 at 21 18 11

File: PlayCover/Utils/NavigationStack.swift [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+//  NavigationStack.swift
+//  PlayCover
+//
+//  Created by Amir Mohammadi on 11/2/1401 AP.
+//
+
+import SwiftUI
+
+enum StackNavigationTransition: Equatable {
+    case none, defaultTranisition, custom(AnyTransition)
```
Comment (JoseMoreville) [inline PlayCover/Utils/NavigationStack.swift]: There is a typo here

File: PlayCover/Utils/NavigationStack.swift [hunk 2]
```diff
@@ -0,0 +1,111 @@
+//
+//  NavigationStack.swift
+//  PlayCover
+//
+//  Created by Amir Mohammadi on 11/2/1401 AP.
+//
+
+import SwiftUI
+
+enum StackNavigationTransition: Equatable {
```
Comment (JoseMoreville) [inline PlayCover/Utils/NavigationStack.swift:10]: Is Equatable the best option to handle this operations?

File: PlayCover/ViewModel/NetworkVM.swift [hunk 3]
```diff
@@ -7,6 +7,7 @@
 
 import SystemConfiguration
 import Foundation
+import Semaphore
```
Comment (JoseMoreville) [inline PlayCover/ViewModel/NetworkVM.swift]: Wasn't `import Semaphore` added before? I'm pretty much sure DispatchSemaphore can handle it without explicitly importing Semaphore

File: PlayCover/ViewModel/StoreVM.swift [hunk 4]
```diff
@@ -6,184 +6,192 @@
...
             }
         }
     }
+}
 
-    func appendSourceData(_ data: SourceData) {
-        self.sources.append(data)
-        self.resolveSources()
-    }
+// Source Data Structure
```
Comment (JoseMoreville) [inline PlayCover/ViewModel/StoreVM.swift:174]: These structs would be better on a different file

File: PlayCover/ViewModel/StoreVM.swift [hunk 5]
```diff
@@ -6,184 +6,192 @@
...
-                if index > self.sources.endIndex {
-                    index = self.sources.endIndex
+    //
+    private func getSourceData(sourceLink: String) async -> (SourceJSON?, SourceValidation) {
+        guard let url = URL(string: sourceLink) else { return (nil, .badurl) }
+        do {
+            let (data, response) = try await URLSession.shared.data(
+                for: URLRequest(url: url, cachePolicy: .reloadIgnoringLocalAndRemoteCacheData)
+            )
+            if !url.isFileURL {
```
Comment (JoseMoreville) [inline PlayCover/ViewModel/StoreVM.swift:147]: You should check for local files before starting any URLRequest

File: PlayCover/Views/Sidebar Views/IPALibraryView.swift
Comment (JoseMoreville) [inline PlayCover/Views/Sidebar Views/IPALibraryView.swift:1]: Can you please separate Sources view(Source Selection view) and Source library view (Apps of a source)

**Review by JoseMoreville — CHANGES_REQUESTED** [review]
Left some comments that need to be checked

File: PlayCover/ViewModel/StoreVM.swift (hunk 5 above)
Comment (amirsaam) [reply PlayCover/ViewModel/StoreVM.swift:147]: Even if URL is local file we need to do URLRequest (or I am worng?)
So we are just checking afterwards to see if the url is local file and bypass the 200 statuscode check for it. nothing more

File: PlayCover/ViewModel/NetworkVM.swift (hunk 3 above)
Comment (amirsaam) [reply PlayCover/ViewModel/NetworkVM.swift]: Actually that is a mistaken import there, DispatchSemaphore is still in authority to check the NetworkVM not AsyncSemaphore. I'll remove.

File: PlayCover/ViewModel/StoreVM.swift (hunk 5 above)
Comment (JoseMoreville) [reply PlayCover/ViewModel/StoreVM.swift:147]: As long it does not depend on response it's ok to move it, but there are better ways to check for local files, although the code works and the order is right, you should check for that before. just a suggestion.

File: PlayCover/Utils/NavigationStack.swift (hunk 2 above)
Comment (amirsaam) [reply PlayCover/Utils/NavigationStack.swift:10]: Actually yes, and those checks were suggested by KhoraLee before here https://github.com/PlayCover/PlayCover/pull/732

File: PlayCover/ViewModel/StoreVM.swift (hunk 5 above)
Comment (amirsaam) [reply PlayCover/ViewModel/StoreVM.swift:147]: surely we need the `response` to be present to bypass it, right?

Comment (amirsaam) [reply PlayCover/ViewModel/StoreVM.swift:147]: looks like it was a misunderstanding. resolving this.

File: PlayCover/ViewModel/StoreVM.swift (hunk 4 above)
Comment (amirsaam) [reply PlayCover/ViewModel/StoreVM.swift:174]: we have no file for data structs, it is somehow redundant to create one just for source structs; isn't it?

**Comment by amirsaam** [comment]
10 month? nice.

**Comment by amirsaam** [comment]
@Depal1 are you interested in this pr? if yes i shall fix its issues, otherwise close it pls.

**Comment by Depal1** [comment]
> @Depal1 are you interested in this pr? if yes i shall fix its issues, otherwise close it pls.

I am, the UI changes sought with this PR would be very useful to manage large IPA collections for testing.

**Comment by amirsaam** [comment]
> > @Depal1 are you interested in this pr? if yes i shall fix its issues, otherwise close it pls.
> 
> I am, the UI changes sought with this PR would be very useful to manage large IPA collections for testing.

Then I try to refactor and rebase it today.

**Pull request closed without merging**